### ISSUE TITLE
Mac OS Classic file loading improvements

### ISSUE THREAD

**Issue by gbody** [uxmal/reko#490]
After having a look at the loaded segments for the program I was looking at , I have the following suggestions. 

1) Exclude the first 4 bytes defining the length of the Resource and just load the byte array for that length into the memory. If the resource is a code resource (with the exception of code resource#0) exclude an additional 4 bytes defining the jumptable offset (word) from the start of the jumptable and the number of entries (word) in the jumptable for that code resource. 

2) I was wondering about how you handle disassembly with respect to the A5 World environment used by the Macintosh. The A5 register is used normally with +(positive) offsets to access a jump table for interseg function calls and -(negative) offsets for access to global variables. Space used for this is usually allocated toward the top of available heap space (upper memory).  See item 3 for setting up

3) Code#0 which contains the information for the jump table and how much space is required below A5 for the global variables should be loaded with enough padding for the global variables , followed by a small offset and finally the jump table. +0 length of resource (Long), +4 Space above A5 for jumptable (Long), +8 Space for global variables (Long), +$C Length of Jumptable (long) +$10 offset from A5 to place Jumptable (Long), then the jumptable data follows.

4) Initial disassembly should start at code resource 1 with the offset into code resource being the first word of the jumptable. In the example program I attached for issue #488, the offset would be $39F0 from the start of CODE:Main#1. 
Code looks like this
﻿0010455A 20 57 movea.l(a7),a0
0010455C 42 A7 clr.l-(a7)
0010455E 48 6D FB 8C pea$-0474(a5)
00104562 48 6D FB 88 pea$-0478(a5)
00104566 48 6D FB 84 pea$-047C(a5)
0010456A 2F 08 move.la0,-(a7)
0010456C 4E BA 02 5E jsr.l001047CC       

The first subroutine starts like this 
﻿001047CC 4E 56 00 00 linka6,#$0000
001047D0 48 E7 13 38 movem.ld3/d6-d7/a2-a4,-(a7)
001047D4 2E 2E 00 18 move.l$0018(a6),d7
001047D8 26 6E 00 0C movea.l$000C(a6),a3
001047DC 4E AD 03 32 jsr.l$0332(a5)

The last instruction shown calls into the jumptable @ offset $332 .

Jumptable @offset $330
﻿00100B5A 00 00                                              ; offset to function
00100B5C 3F 3C 00 03 move.w#$0003,-(a7)  ; code resource 3
00100B60 A9 F0 _LoadSeg                             ; load and execute function

So the interseg function being called is at offset $0000 of code resource 3

5) %A5Init basically clears the global variable space , unpacks the global variable data and does some relocation. Once I get setup I will look to add to the loader to unpack the global data into the space allocated for the A5 world

6) I will look to add additional function to the binhex loader to the decode MACBug function names if available in the binary

**Comment by uxmal** [comment]
@gbody: thanks for taking the time to write the suggestions above. Since I have scant experience programming on classic MacOS, any insights or clarifications you can provide are most appreciated. 

1. Currently, Reko loads the entire resource fork as a big blob into memory. To mimic the way MacOS deals with the segments within the resource fork, perhaps it is more appropriate to read in each resource within the fork as a separate blob? Then, for each loaded resource blob, we can elide the first 4 bytes as these always seem to be the size of the resource, and in the case of `CODE` resources whose number is not 0, the next 4 are also elided.

BTW: the loader is arbitrarily loading the image at address $00100000, for lack of better knowledge. Can you recommend a more appropriate/typical load address?

2. I'm aware of the A5 world from discussions with other MacOS savvy people, but currently no special steps are taken to label the globals there. The MacOSClassic platform would have to inject type information after the resource fork is loaded. I don't have a source for the A5 global variables (names, offsets, data types) -- perhaps you know of a good reference? Is it safe to assume that on the MacOS platform , the `A5` register is always pointing to the A5 world? If so, some creative extensions to the `IPlatform` interface could be made to resolve accesses of the form `move.l $-NNNN(a5),d0` into global memory accesses. I just need those offsets...

3. If I understand you correctly, in addition to the `CODE#0` segment, the Reko loader should create a new segment -- perhaps called `A5world`? -- which contains a "resolved" copy of the jump table in `CODE#0`. That is, all the cross-segment references are replaced with jumps to the intended procedures. Looking at this [reference](http://mirror.informatimago.com/next/developer.apple.com/documentation/mac/Processes/Processes-138.html#HEADING138-0) I see that this segment should have the size of `sizeof(QuickDrawGlobals) + sizeof(ApplicationGlobals) + sizeof(ptrToQuickDrawGlobals) + sizeof(ApplicationParameters) + sizeof(JumpTable)`. It seems space for the QuickDrawGlobals section is probably not needed during decompilation.  I know the size of ApplicationGlobals  and JumpTable from `CODE#0`, and the size of a pointer is 4 bytes. Do I have to do something special with what the documentation calls ApplicationParameters, or do I just leave a hole the size of `A5 offset of jump table` there?

4. Currently Reko is not taking advantage of all those pointers in the jump table. They provide useful "roots" for the procedure detection / scanning phase. It's easy enough to do once the `A5world` segment has been generated.

It sounds like you're in the process of rolling up your sleeves and getting your hands dirty with the Reko codebase. If this is the case, don't hesitate to send me any questions you may have. I'm glad to see interest in the MacOS part of Reko and am keen on improving on its current half-finished state.

**Comment by gbody** [comment]
@uxmal It might take me a little while to get my head around C#, but lets see if I can provide something useful.

I wasn't thinking of both a code0 and A5world segment, but just a single A5world segment. Using the data from Code0 to populate the jumptable. My understanding is that for decompiling, A5 could be considered a fixed value for the application. 
Previously when I mentioned global variables, I should have said Application global variables are referenced using (negative) offsets from A5World. Because it is the application global variables there is no known variable names that can/should be prepared from references.

As for where to place the program in memory, the original Mac has like 128KB and that was the main reason for the segmentation, allowing larger programs to run in the small memory space. I don't think it will matter to much loading things at $100000 for the decompiling.

Is there currently a way to handle library functions that may have been linked in the application code, so library functions can be excluded from the code and reduced to a function name. eg. printf

**Comment by uxmal** [comment]
@gbody: I've completed most of the tasks above on my machine. My final push is contingent on fixing a regression I uncovered as part of the work. I decided to keep both the CODE#0 segment and the A5World segments, to allow a user to see the segments as they are on disk as well as the "live" version.  Once you get a chance to play with it you can advise me on whether this is redundant or not.

As I understood it from the documentation, not only the application globals are stored in the "negative side" of A5World, but also Toolbox globals. It's those globals I would like to have references for.

I will leave the base loading address as-is, per your advice above.

Unfortunately, Reko currently has no automated way to hide library functions that are linked into binaries. It is a frequently requested feature that I hope to start working on after some major core decompiler tasks have been dispatched. In the meanwhile, you can mark individual procedures as "no-decompile" by right-clicking on them in the user interface and selecting "Properties". In the dialog you can override the procedure name and indicate whether the decompiler should process the body of the procedure or not. If the latter, you will be required to specify the signature of the procedure in an augmented C syntax. 

The augmentation revolves around the mechanism used to denote registers. This is important in platforms where there is not defined standard calling conventions, or in the case of procedures that violate the ABI of a particular platform. In such cases, you would specify the signature of a procedure like this:
```C++
[[reko::reg("a3")]] return_pointer_in_a3([[reko::reg("a2")]] arg_in_a2);
```
Another thing I just realized is that some of the toolbox APIs have a really weird return value convention. Instead of returning results in registers, they actually place the result on the stack. Reko will need some changes to implement this. Fortunately, in the x86 world, a procedure that returns a value using floating point numbers will leave the value on the Intel x87 processor stack, so it should be fairly easy to adapt that code to the m68k.

Having code to parse those MACBug symbols will be nice. If you find C# to be too daunting, you could always provide me with the equivalent code in your preferred language; I'd be happy to translate it for you.

**Comment by uxmal** [comment]
@gbody: calls of the form `jsr $offset(a5)` are now resolved into direct calls to the target procedures. Looks very good in the sample you provided me.

Do you know where one can find a detailed specification of the Toolbox system calls, paying attention to what registers/stack locations are used for parameters and return values? As long as it's a text file I can convert it to a format that Reko can use.

**Comment by gbody** [comment]
@uxmal Your been doing some nice work on this.
 
I don't know if someone has done anything like thatfor the  Toolbox system calls, as I haven't spent a huge amount of time developing apps for the Macintosh. Looking at inside Macintosh vol I - III chapter 4 should give you a decent idea as a starting point.  It was continually updated over the years.
http://www.weihenstephan.org/~michaste/pagetable/mac/Inside_Macintosh.pdf
I will ask a few people I know that have plenty years of Mac development behind them if they are aware of anything like that.

I still think CODE#0 does not need to be added as a separate resource.

I have been able to build your new version and the Interseg jumps look good.

1)There is an issue with the Main program loop of the application at 100B6A. The initial "scan binary"  has this routine starting and ending and new function starting at 100B6E.  

2)Failing Resource internal jump tables failing.
Warning,001066A4,The call or jump table has invalid addresses; stopping.
Warning,001066A0,No valid entries could be found in jump vector.
Warning,00101CA0,The call or jump table has invalid addresses; stopping.
Warning,00101C9C,No valid entries could be found in jump vector.

There are incomplete sections that are due to these failing jump tables. These tables are made up of Word offsets and I have attached a file with some code examples from the same file. The code at these jumptable offsets would have accessed a lot of the additional subroutines filling in the incomplete sections. 
 
[MacOS resource internal jump table issue.txt](https://github.com/uxmal/reko/files/1430805/MacOS.resource.internal.jump.table.issue.txt)

3) the Ver#2 resource causes an exception error when selecting.

At this stage the loading only needs to have the application globals unpacked and MacsBug function names if available applied. 
Reference for the function names is from MacsBug Reference 6.2 Appendix D. http://bebop.gtxent.com/MacsBugRef_6.2.pdf

**Comment by uxmal** [comment]
I will remove CODE#0 per your request.

I read MacsBug Appendix D and will use a crude algorithm like the one described there to extract symbolic information.

The m68k relative jump table issue is known (see #121, #122 for past discussions). I will take a look to see if there is anything I can do in the short term. Long term, issue #132 tracks the general problem.

I will also look at the main program loop to see why the first word is being chopped off.

**Comment by gbody** [comment]
@uxmal  is the backwalker.cs file being used to try and identify the table size. Also is it only used for 68K Arch.

**Comment by uxmal** [comment]
It is used to identify table sizes and it is cross platform: it works quite well for Windows or MS-DOS programs where the table contains absolute, pointer-sized addresses.

**Comment by gbody** [comment]
@uxmal This link might be more what your looking for as far as globals, error values extra.
http://www.mac.linux-m68k.org/devel/macalmanac.php


How would you like the calls encoded for importing

Hear is an example of a Register based call 
_DisposHandle  in A0 Handle (4 bytes), out D0 Result (Word)

Example of of stack based call
FUNCTION GrowWindow (theWindow: WindowPtr; startPt: Point; sizeRect: Rect) LONGINT; 

SUBQ.L #4,SP                    ;make room for LONGINT result
MOVE.L theWindow,-(SP) ;push window pointer
MOVE.L startPt,-(SP)  ;a Point is a 4-byte record,
                                  ; so push actual contents
PEA sizeRect              ;a Rect is an 8-byte record,
                                  ; so push a pointer to it
_GrowWindow          ;trap to routine
MOVE.L (SP)+,D3      ;pop result from stack 


list of traps
[Mac traps list.txt](https://github.com/uxmal/reko/files/1437894/Mac.traps.list.txt)

**Comment by uxmal** [comment]
Thanks for the global references! Those can be made into a memory map which Reko can use to identify global vars. 

As for the syscalls, there are two choices. Initially, Reko used an XML-based format to encode all system calls. For MacOS, these can be found in the file `$(REKO)/src/Environments/MacOS/macos_classic.xml`. A typical register-based call looks like this today:
```xml
  <service name="_DisposHandle"><syscallinfo><vector>A023</vector></syscallinfo>
    <signature>
      <arg name="handle"><prim domain="UnsignedInt" size="4" /><reg>a0</reg></arg>
      <return><prim domain="UnsignedInt" size="2" /><reg>d0</reg></return>
    </signature>
  </service>
```
However, this is very tedious to write. Recently we added support in Reko's C parser to handle C++ attributes. These can be used to denote information that isn't part of the C language (like registers used in a call. It shouldn't be too hard to extend the attribute logic to fully replace the XML representation. The equivalent of the XML above could be:
```C++
[[reko::syscall(0xA023)]]
[[reko::reg("d0")]] uint16_t _DisposHandle([[reko::reg("a0")]] Handle h);
```
..which is arguably slightly more legible than the XML.

The stack-based calls are probably best implemented by defining a new MacOS-specific calling convention (say: "__stack_based"?) and then declaring the function as:
```C++
long [[reko::convention("stack_based")]] GrowWindow(
    WindowPtr theWindow, Point startPt, Rect sizeRect);
```
Implementing a calling convention for stack-based calls is not terribly hard. I will open a separate issue to track this.

However, none of the references you provided contain information about how parameters are passed, be it via registers or the stack. It would be very useful to have that information in some machine readable format. Then, it could be transformed into the C++ syntax described above and committed into the project.

**Comment by uxmal** [comment]
@gbody : I was able to locate a PDF file online that seems to contain a large percentage of all the MacOS Classic toolbox calls along with other services. I extracted as much text as I could, then wrote a Pascal Parser to extract all the `INLINE` + m68k opcode sequences. These are used to extract register and stack usage signatures. There is therefore no need to write any other form of metadata file!

Regrettably the PDF file isn't 100% complete, so a mild regression was taken: some of the services previously detected in the sample binary were lost. It shouldn't be too hard to fix the remaining ones. There are 71 locations with undetected services, and many of them are duplicates. 

Once I've finished those, I think I will declare this issue as addresses. I can then look at the other issues you reported.

**Comment by gbody** [comment]
@uxmal I can provide the interface files from MPW from 2001 as either aincludes, cincludes or pincludes. I had only just started working through the system A-traps.

**Comment by uxmal** [comment]
I have resolved all the A-traps except for $A9EB (`_FP68K`). This trap seems to provide IEEE floating point "sub-services", by pushing parameters and a function selector onto the stack before executing the trap instructions.  My google-fu is too weak to find the 68k documentation for Apple's SANE library. Perhaps you have the necessary docs? FWIW, in your binary, the only call to `_FP68K` is passing function selector 0x000B.

**Comment by uxmal** [comment]
@gbody: I consider is issue address. Minus the FP68k support mentioned above -- which really should be a separate issue -- Reko now loads the BinHex-encoded binary fine, including resolving the cross-segment calls in a sensible way. I leave it to you to close this issue. If you are able to provide documentation about _FP68k, please either submit a separate issue along with a pointer to said documentation, or if you're feeling very brave, submit a pull request :-)

**Comment by gbody** [comment]
@uxmal the macintosh almanac has a section _PACK 4 that is the same trap number. The listed sub functions are listed there. The high byte definitions for data type aren't there though. Once I have the numerics info available, I will raise a new case. Happy to close this now.
